google/cloud/bigtable_v2/services/bigtable/transports/pooled_grpc_asyncio.py

```diff
@@ -0,0 +1,597 @@
+        for i in range(pool_size):
+            new_channel = type(self).create_channel(
+                self._host,
+                # use the credentials which are saved
+                credentials=self._credentials,
+                # Set ``credentials_file`` to ``None`` here as
+                # the credentials that we saved earlier should be used.
+                credentials_file=None,
+                scopes=self._scopes,
+                ssl_credentials=self._ssl_channel_credentials,
+                quota_project_id=self._quota_project_id,
+                options=[
+                    ("grpc.max_send_message_length", -1),
+                    ("grpc.max_receive_message_length", -1),
+                ],
+            )
+            self.channel_pool.append(new_channel)
+
+        # Wrap messages. This must be done after self.channel_pool is populated
+        self._prep_wrapped_messages(client_info)
+
+    def next_channel(self) -> aio.Channel:
+        """Returns the next channel in the round robin pool."""
+        # Return the channel from cache.
+        channel = self.channel_pool[self._next_idx]
+        self._next_idx = (self._next_idx + 1) % len(self.channel_pool)
+        return channel
+
+    async def replace_channel(
+        self, channel_idx, grace=None, new_channel=None
+    ) -> aio.Channel:
+        """
+        Replaces a channel in the pool with a fresh one.
+
+        The `new_channel` will start processing new requests immidiately,
+        but the old channel will continue serving existing clients for `grace` seconds
+
+        Args:
+          channel_idx(int): the channel index in the pool to replace
+          grace(Optional[float]): The time to wait until all active RPCs are
+            finished. If a grace period is not specified (by passing None for
+            grace), all existing RPCs are cancelled immediately.
+          new_channel(grpc.aio.Channel): a new channel to insert into the pool
+            at `channel_idx`. If `None`, a new channel will be created.
+        """
+        if channel_idx >= len(self.channel_pool) or channel_idx < 0:
+            raise ValueError(
+                f"invalid channel_idx {channel_idx} for pool size {len(self.channel_pool)}"
+            )
+        if new_channel is None:
+            new_channel = self.create_channel(
+                self._host,
+                credentials=self._credentials,
+                credentials_file=None,
+                scopes=self._scopes,
+                ssl_credentials=self._ssl_channel_credentials,
+                quota_project_id=self._quota_project_id,
+                options=[
+                    ("grpc.max_send_message_length", -1),
+                    ("grpc.max_receive_message_length", -1),
+                ],
+            )
+        old_channel = self.channel_pool[channel_idx]
+        self.channel_pool[channel_idx] = new_channel
+        await old_channel.close(grace=grace)
```

Comment:
Not sure about python threading model, but I think you have a race condition:

close() even with grace will stop accepting new rpcs.
Can this happen?
- caller thread calls next_channel() and grabs a reference channel 1
- refresh thread calls replace_channel() and close() on channel 1
- caller thread tries to start a call using unary_stream() on channel1

Reply:
In general, we shouldn't have to worry about this as much here as in other languages, because [async grpc channels can't be shared from multiple threads](https://grpc.github.io/grpc/python/grpc_asyncio.html#caveats), and asyncio is pretty particular about when context switches happen

That said, this is a good point, and it's probably a good idea to take some precautions in case those assumptions change. I'll add an extra sleep buffer in between swapping channels and closing the old one, and I'll look into if there are any good places to set locks to be safe